ramips: add support for Tenda A6 (WL0175, RT5350)

fragtion opened the pull request:
Hi

Small patch to include support for Tenda A6 mini wireless router

Specifications are already detailed at https://openwrt.org/toh/tenda/a6

I have published additional details at following custom build github page, which I am using/offering until device support is official: https://github.com/fragtion/openwrt-hardware/tree/master/tenda-a6

As my builds are working very well for me, it would be nice to make this more widely available/accessible to the general public

Thank you for your consideration :)

Signed-off-by: Dimitri Pappas <fragtion@gmail.com>

fragtion committed:
ramips: add support for Tenda A6 (WL0175, RT5350)

Tenda A6 is a miniature portable indoor wireless AP for 2.4GHz with single Ethernet port, based on Ralink/MediaTek RT5350F soc. It is very similar to the unbranded A5-V11.

Device specifications:
- SoC: RT5350F
- CPU Frequency: 360 MHz
- Flash Chip: Macronix MX25L3206E (4096 KiB)
- RAM: EtronTech EM63A165TS (32768 KiB)
- 1x 10/100 Mbps Ethernet (4x LAN, 1x WAN)
- 1x internal 2.4GHz antenna (non-PCB)
- UART (TP1 & TP2) header on PCB (57600 8N1)
- Wireless: SoC-intergated: 2.4GHz 802.11bgn
- USB: None populated (unsoldered contacts: J100)
- 1x LED, 1x reset button

Flashing instructions:
1. Assuming router is on factory IP (192.168.2.1/24), configure PC with static IP between 192.168.2.2/24 and 192.168.2.254/24
2. Access router's web UI at http://192.168.2.1, and proceed to navigate to firmware update page
3. Flash the (temporary) initramfs version of OpenWrt image "openwrt-ramips-rt305x-a6-initramfs-kernel.bin"
4. Allow sufficient time (~5 minutes) for upgrade process to complete. During this waiting time, proceed to configure new static IP between 192.168.1.2/24 and 192.168.1.254/24
5. Using SCP, transfer "openwrt-ramips-rt305x-a6-squashfs-sysupgrade.bin" to root@192.168.1.1:/tmp/ (alternatively, SSH to root@192.168.1.1, and then wget that image from HTTP server on same subnet)
6. Using SSH, log into root@192.168.1.1, and execute "sysupgrade -F -n /tmp/openwrt-ramips-rt305x-a6-squashfs-sysupgrade.bin"
7. Device will self-reboot again, and take another ~5 minutes to upgrade & boot into working instance of OpenWrt-squashfs.

Signed-off-by: Dimitri Pappas <fragtion@gmail.com> (github: fragtion)

adrianschmutzler commented:
Hi, your commit has formal issues, please visit https://openwrt.org/submitting-patches

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+/dts-v1/;
```
Please add an SPDX license identifier.

fragtion replied:
Hrm, where do I find that?

adrianschmutzler replied:
In all recently (at least for the last year) added DTS files.

adrianschmutzler replied:
https://spdx.org/licenses/

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+
+/ {
+	compatible = "tenda,a6", "ralink,rt5350-soc";
+	model = "A6";
```
Tenda A6?

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+		compatible = "gpio-leds";
+
+		led_power: power {
+			label = "tenda:blue:ld1";
```
a6:blue:power?

fragtion replied:
Affirmative

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
flash@0

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
faster speed possible?

fragtion replied:
How can i check? this works fine

adrianschmutzler replied:
Checking data and testing would be nice. Typically, faster values are possible and yield improvement, but just everyone copies the 10 MHz at first.

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+
+&state_default {
+	gpio {
+		groups = "i2c", "jtag", "uartf", "led";
```
do you need them all?

fragtion replied:
I may not, but it's probably best practice to include for wider potential use. Size impact should be neglibible?

adrianschmutzler replied:
It's common practice to include what's needed.

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x4>;
```
Is there an address in 0x28, 0x2e, 0xe000 or 0xe006?

fragtion replied:
Eh, I'll see what I can do, this may be a bit out of my reach as I'm not too clued up about address registers or hex codes

adrianschmutzler replied:
https://openwrt.org/docs/guide-developer/mac.address#find_out_about_flash_locations

Essentially, it's just doing a hexdump and looking where MAC address are found in the factory partition.

How are the addresses assigned at vendor firmware? Based on your current settings, WiFi and ethernet use the same address (and that one is also on the label)?

adrianschmutzler commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,111 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
```
Use 0x0 for consistency.

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1086,6 +1086,20 @@ define Device/unbranded_a5-v11
 endef
 TARGET_DEVICES += unbranded_a5-v11
 
+define Device/tenda_a6
```
Please relocate due to alphabetic sorting.

fragtion replied:
Roger

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1086,6 +1086,20 @@ define Device/unbranded_a5-v11
+  DEVICE_VENDOR := Tenda
+  DEVICE_MODEL := A6
+  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2
+  SUPPORTED_DEVICES += a6
```
This line can be dropped.

fragtion replied:
Which one? You quoted 4. The device has an unpopulated USB header, but the SoC has working USB support, so it's probably best to include the modules

adrianschmutzler replied:
The line directly before the comment.

SUPPORTED_DEVICES in set automatically based on the device node name. Despite, the value you provide does not match anything anyway.

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1086,6 +1086,20 @@ define Device/unbranded_a5-v11
+  DEVICE_MODEL := A6
+  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2
+  SUPPORTED_DEVICES += a6
+  DEFAULT := n
```
Either the device works with default settings when added or we won't accept it.

adrianschmutzler commented on target/linux/ramips/image/rt305x.mk:
```diff
@@ -1086,6 +1086,20 @@ define Device/unbranded_a5-v11
+  SUPPORTED_DEVICES += a6
+  DEFAULT := n
+endef
+TARGET_DEVICES += unbranded_a5-v11
```
Copy-paste error. So, did you even test it?

fragtion replied:
Everything was working on tarball-based build, I had to redo it for the github master branch in order to submit the patch. I quickly posted a follow-up commit to address this error, but yes as you have noted there are some further inconsistencies like the missing entry in 02_network. As far as marking 0x0 in the DTS file, I feel inclined to argue that you are being overly pedantic, as I used the accepted `unbranded_a5-v11` as base template, which according to you would be erroneous as well. You may want to check that out :) Regards

adrianschmutzler replied:
The 0x0 is entirely cosmetic, it will work as well as just "0", but I prefer the former as it will match the other locations which are typically in hex notation. However, this is the least important of all request IIRC.
Despite, if the improvements we could ask for would always be limited to what somebody accepted for some devices in the past, there would never be improvement. I give feedback based on what I think would be best for supporting the particular device at hand.

fragtion replied:
You are right. Look, I concede that this is mostly a copy-paste operation for the most part, but it's what was largely the basis of what was required to get this device running OpenWrt at all, if that makes sense. Appreciate all of your time and will try be more careful and compliant for the next pull, if I find the time to make the requested adjustments, both of the necessary and cosmetic kinds. Take care :D

adrianschmutzler commented:
no entry in 02_network?

adrianschmutzler commented:
Please merge all your changes into a single, proper commit.

fragtion commented:
Thanks for the thorough inspection, I shall rehash and resubmit the pull !

adrianschmutzler commented:
> Appreciate all of your time and will try be more careful and compliant for the next pull, if I find the time to make the requested adjustments, both of the necessary and cosmetic kinds.

Thanks. Please make sure to force-push to the existing branch and update this PR, please don't open a new PR for the changed commit.

And of course, the most important question will be whether this can actually be built with buildbot settings (keeping the DEFAULT := n in mind), because otherwise you would be wasting your time.

fragtion commented:
Hi Adrian. Squashed and force pushed the corrections for your review.

Some notes on the changes:

`target/linux/ramips/dts/rt5350_tenda_a6.dts`:
According to datasheet for RT5350, SPI_CLK pin/XTAL_FREQ_HI (I'm assuming this is the same frequency item we are concerned with) is 20MHz by default, but supports 40MHz also. Since DTS file specifically wants spi-**max**-frequency, I have adjusted value to <40000000> accordingly. 
_For reference, the datasheet is at https://cdn.sparkfun.com/datasheets/Wireless/WiFi/RT5350.pdf_

I could also confirm that there is indeed only a single MAC address in Tenda's/factory 4MiB firmware , namely at Offset 262,144 (0x40000), the only match for C8:3A:35 (MAC prefix for Tenda Technology Co., Ltd.). The full MAC also correlates with the single MAC address printed on the sticker underneath my router, so we can be sure of it

For the GPIO groups, I see that the .dts files for all existing RT5350-based devices share at least "i2c", "jtag", and "uartf" in common (with the only exception being "led" in some cases). While we obviously shouldn't presumptuously just base this new device on previous consensus patterns, I can confirm that this device does have a LED (&gpio0 7, as defined in dts file) - so "led" is needed / should be kept. "i2c" relates to the EEPROM as far as I can see, and "jtag" & "uartf" are both documented in the schematic and can be useful for debugging (I personally used uart console for debugging purposes) - so again it is advisable to keep them both, too. My suggestion is therefore to leave that line unaltered

Trusting it looks much better to you now. I have built it and tested the images and it is working. Users must just be aware that this is one of those cases where they cannot flash directly from factory firmware to squashfs but must go through initramfs image and manual sysupgrade process via cli.

Apologies again for the mess-ups earlier / thank you for your patience - it was very reckless of me to post the pull request without doing a thorough test on that specific code and will make every effort to avoid that again.  I hope to make more/further contributions to this great project in future and that it will continue to evolve and improve for years more to come.

Best Regards

adrianschmutzler commented:
> I could also confirm that there is indeed only a single MAC address in Tenda's/factory 4MiB firmware , namely at Offset 262,144 (0x40000), the only match for C8:3A:35 (MAC prefix for Tenda Technology Co., Ltd.).

If it's on 0x40000, then the current setup is wrong. Did you mean 0x40004?

adrianschmutzler commented on target/linux/ramips/rt305x/base-files/etc/board.d/02_network:
```diff
@@ -207,6 +208,7 @@ ramips_setup_macs()
 	omnima,miniembplug|\
 	planex,mzk-w300nh2|\
 	sitecom,wl-351|\
+	tenda,a6|\
```
Above, you define it as a one-port. So why should it have an address for WAN defined?

adrianschmutzler commented:
Hi, you still didn't address the formal issues, i.e.

- commit message (specs, installation)
- Signed-off-by

Despite, please make the "add" in the commit title lower-case

fragtion commented:
> > I could also confirm that there is indeed only a single MAC address in Tenda's/factory 4MiB firmware , namely at Offset 262,144 (0x40000), the only match for C8:3A:35 (MAC prefix for Tenda Technology Co., Ltd.).
> 
> If it's on 0x40000, then the current setup is wrong. Did you mean 0x40004?

Yes, sorry. it is 0x40004 to be precise. I thought there was some important "init" code at 0x40000 which should be included as well, but the MAC itself starts at 0x40004 so that is another error I will correct.

> Above, you define it as a one-port. So why should it have an address for WAN defined?

It is defined to use the same MAC (factory 0x4) as ethernet as you can see. I was unsure whether or not that should be explicitly defined. Given how strict this process seems to be, I thought to rather be on the safe side by doing so rather than not. 

I shall attend to specs & installation details in some hours if I can (after reading through https://openwrt.org/submitting-patches, to make sure I am not missing anything and causing more tedious back-and-forth over here), and then sign off (again?) as required. Thanks chief

adrianschmutzler commented:
> Yes, sorry. it is 0x40004 to be precise. I thought there was some important "init" code at 0x40000 which should be included as well, but the MAC itself starts at 0x40004 so that is another error I will correct.

There is some "init code" at 0x0 of the factory partition (your 0x40000), the calibration data of the wmac (if you set it up correctly). But this discussion was about the mac addresses, so I was expecting their locations.

adrianschmutzler commented:
> but the MAC itself starts at 0x40004 so that is another error I will correct.

What error?

adrianschmutzler commented:
>> Above, you define it as a one-port. So why should it have an address for WAN defined?

> It is defined to use the same MAC (factory 0x4) as ethernet as you can see. I was unsure whether or not that should be explicitly defined. Given how strict this process seems to be, I thought to rather be on the safe side by doing so rather than not.

No, you defined to take the stored MAC address, add 1 to it, and then apply this address to the non-existant WAN interface (!= what you set with &ethernet in DT). Anyway, now you removed that step.

So as you see, this is not about being pedantic, but about finding out where you put values/code you didn't really understand in the first place.

adrianschmutzler commented:
> and then sign off (again?) as required

Just found that you put a Signed-off in the first comment. It needs to be in the commit message. That's what will be pushed to the repository after all.

adrianschmutzler commented:
Despite, take your time, device support PRs typically take some time until they are reviewed and merged.

mans0n commented:
Hi, I'll explain more for the gpio pinctrl. The purpose of the node is for enabling GPIO instead of other features such as I2C and JTAG. The strings specify which pin groups we are going to use. (Please refer page 12 of the datasheet.) If we add "i2c" to the node, it means we disable i2c and use the pins (i.e. gpio1 and gpio2) for gpio. If we want some hardware features to be enabled (like jtag), then we should _not_ add the pin groups to the gpio node.
In this case, only gpio0 and gpio7 are used, so we only need "uartf". Other groups can be safely removed.

mans0n commented on target/linux/ramips/dts/rt5350_tenda_a6.dts:
```diff
@@ -0,0 +1,112 @@
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
Please try "gpio-keys" here. `poll-interval` can be removed then.

adrianschmutzler replied:
@mans0n Is that supported for the old subtargets?

When grepping, I did not find a single case of "gpio-keys" for any rt**** devices.

mans0n replied:
Probably. At least I've tested it on cy-swr1100.

adrianschmutzler commented:
@fragtion Ping.

fragtion commented:
Device specifications:
* SoC: RT5350F
* CPU Frequency: 360 MHz
* Flash Chip: Macronix MX25L3206E (4096 KiB)
* RAM: EtronTech EM63A165TS (32768 KiB)
* 1x 10/100 Mbps Ethernet (4x LAN, 1x WAN)
* 1x internal 2.4GHz antenna (non-PCB)
* UART (TP1 & TP2) header on PCB (57600 8N1)
* Wireless: SoC-intergated: 2.4GHz 802.11bgn
* USB: None populated (unsoldered contacts: J100)
* 1x LED, 1x reset button

Flashing instructions:
1. Assuming router is on factory IP (192.168.2.1/24), configure PC with static IP between 192.168.2.2/24 and 192.168.2.254/24
2. Access router's web UI at http://192.168.2.1, and proceed to navigate to firmware update page
3. Flash the (temporary) initramfs version of OpenWrt image "openwrt-ramips-rt305x-a6-initramfs-kernel.bin"
4. Allow sufficient time (~5 minutes) for upgrade process to complete. During this waiting time, proceed to configure new static IP between 192.168.1.2/24 and 192.168.1.254/24
5. Using SCP, transfer "openwrt-ramips-rt305x-a6-squashfs-sysupgrade.bin" to root@192.168.1.1:/tmp/ (alternatively, SSH to root@192.168.1.1, and then wget that image from HTTP server on same subnet)
6. Using SSH, log into root@192.168.1.1, and execute "sysupgrade -F -n /tmp/openwrt-ramips-rt305x-a6-squashfs-sysupgrade.bin"
7. Device will self-reboot again, and take another ~5 minutes to upgrade & boot into working instance of OpenWrt-squashfs.

Signed-off-by: Dimitri Pappas [fragtion@gmail.com](mailto:fragtion@gmail.com) (github: fragtion)

adschm commented:
Hi, sorry, but I will close now due to the following reasons:
- I'm the only reviewer
- This is a 4/32 device (https://openwrt.org/supported_devices/432_warning)
- There is still work to do (rebase to recent changes, formatting, ...)
- The last change is from 8 months ago

Sorry again, but we cannot use our limited time to review and add devices that already won't work properly anymore when released (in stable) for the first time. This is _not_ your fault, but probably the reason why nobody else cared about this PR already when you posted it.

fragtion commented:
No problem. It's running great on my device... and I've played my part trying to contribute to the project. I have to say, the contribution process is overly convolute and stringent. The OpenWRT team (ie, including yourself) should be doing more to assist with this process, rather than be so critical as is evident by this thread. It's strange that you pinged/rushed me to sign off my changes, but it takes you 8 months to come back and reject them again, and apparently that lapsed time is a reason that counts against me? Lol, anyways one less supported device that could have been supported with minimal input from your side, instead of assisting with the finishing touches to the code, you chose to articulate several lines of reasons why it will now not be included. Hopefully someone else will have the will & patience to revive this. Thanks for your review and oversight/managing of this attempted contribution, and good luck to the future of this otherwise promising project further.

adschm commented:
Keep in mind that we do not assign people to PRs, so the fact that I gave some comments at the beginning does not mean that I'm responsible for this PR. I frequently provide quick feedback for certain areas of submissions, and then leave it for other reviewers to take care of the rest (e.g. because there are parts I cannot review). Unfortunately, I've made the experience that quick responses are actually bad, as people expect the review process to go on like that, or are disappointed if other reviewers need longer. But still, I don't think not providing feedback would be a better procedure here?

However, I should have made a specific comment about the 4/32 situation earlier here, which I seem to have missed...

fragtion commented:
Interesting point, noted. In that case I can only reiterate my gratitude for the assistance you willingly provided and agree that we chalk this off for now, for what it is. I presume that any demand for wanting OpenWrt on the Tenda A6 is probably miniscule, to the point that including support for it in official builds is not worth the added bloat to those images. If anyone truly wants to run it, the work has basically been done so there are sufficient resources to accomplish that by building from source. Good enough in my books anyway. Thanks again & Take care :)

adschm commented:
> I presume that any demand for wanting OpenWrt on the Tenda A6 is probably miniscule, to the point that including support for it in official builds is not worth the added bloat to those images. 

Be aware that we do not provide any images for 4M flash devices anymore (for several months already, actually; just 19.07 is still having them). So, there would not have been official _builds_ anyway, just official _code_.
